Updating CardBlock

vossisboss opened the pull request:
This PR addresses the CardBlock component of #38 

I updated `CardBlock` and added templates for the individual cards as well as the overall card section. This code provides a basic three column design for cards.

Any suggestions for better design or better validation would be appreciated. I wasn't sure if there was a good way to enforce using multiples of three through validation.

laymonage commented on templates/components/streamfield/blocks/card_section_block.html:
```diff
@@ -0,0 +1,15 @@
+<div class="flex flex-col md:flex-row gap-20">
+    {% for card in value.cards %}
+        <div class="flex flex-col gap-2 items-center">
+            {% include "components/streamfield/blocks/card_block.html" with value=card classes="pb-[50px] md:pb-[100px] md:odd:pr-[var(--gutter-xl)] md:even:pl-[var(--gutter-xl)] lg:odd:col-span-2 lg:even:col-span-2 lg:odd:col-start-2 lg:even:col-start-4" %}
```
These `classes` don't seem to be used in `card_block.html`? There's no `{{ classes }}` in there.

Also, I don't see `grid` being used anywhere in `card_block.html`/`card_section_block.html`, so I'm not sure how the `col-start-*` and `col-span-*` classes were supposed to work.

I also don't know where the `gutter-xl` variable comes from, it's not in this project and I don't think it's in Tailwind... Looking at its usage with odd/even padding right/left, would `gap` do the job?

vossisboss replied:
This was definitely a case of me trusting Claude a bit too much. I'll do a better job next time of seeing which classes are ultimately necessary.

vossisboss commented on project_name/utils/blocks.py:
```diff
@@ -107,20 +107,13 @@ class Meta:
     class Meta:
-        icon = "link"
-        value_class = CardStructValue
+        icon = "address-card"
```
From the prior author's imagination I guess. I'll pick something out from our default list instead.

vossisboss commented on templates/components/streamfield/blocks/card_block.html:
```diff
@@ -0,0 +1,17 @@
+    {% endif %}
+    {% if value.link %}
+      {% with cta_link=value.link.0 %}
+        {% include "components/button.html" with title=cta_link.value.get_title url=cta_link.value.get_url arrow=True classes="w-full md:w-auto" %}
```
I am honestly not sure where that came from. I might have been experimenting with adding an icon to the button and forgot to remove it. I'll take it out.

vossisboss commented on project_name/utils/blocks.py:
```diff
@@ -107,20 +107,13 @@ class Meta:
+    link = LinkStreamBlock(required=False, min_num=0)
     class Meta:
-        icon = "link"
-        value_class = CardStructValue
```
One other question: When you have blocks building on top of a base block like `BaseCardSectionBlock` and `CardSectionBlock`, would you ever want to use `StructValue` on them too? I guess it depends on how the templates are structured? Because the corresponding template for `CardSectionBlock` doesn't pull in fields from those cards in a direct way.

laymonage replied:
I'm not sure I follow. You can use a custom `StructValue` for the `BaseCardSectionBlock`/`CardSectionBlock` if you want to add helper methods when dealing with its value.

> Because the corresponding template for `CardSectionBlock` doesn't pull in fields from those cards in a direct way.

I'm not sure what you mean by "pull in fields from those cards in a direct way", but if we use `{% include_block card %}` to render each card in `CardSectionBlock`'s template, we'll use the template defined in `CardBlock` (so they'll be rendered the same way as if you're rendering `CardBlock` individually).

vossisboss commented:
This is ready for you again @laymonage. I poked at the Tailwind classes more thoroughly and removed a `<div>` that didn't seem to be doing much.

I noticed the import statements need cleaning up as well but I don't think we've added linters to this project yet, so perhaps we should make that a clean up a separate PR.

laymonage commented on templates/components/streamfield/blocks/card_section_block.html:
```diff
@@ -0,0 +1,10 @@
+
+<div class="flex flex-col md:flex-row gap-20">
+    {% for card in value.cards %}
+        {% include "components/streamfield/blocks/card_block.html" with value=card %}
```
We can use `{% include_block %}` so we don't have to specify the template again

```suggestion
        {% include_block card %}
```

laymonage commented on templates/components/streamfield/blocks/card_block.html:
```diff
@@ -0,0 +1,17 @@
+      <h3 class="mb-8 text-xl md:text-2xl text-mackerel-300 font-semibold">{{ value.heading }}</h3>
+    {% endif %}
+    {% if value.description %}
+      <div class="mb-8 md:mb-12 md:pb-16 text-lg md:text-xl">
```
Some of the classes in this PR (not sure which ones) don't yet exist in the generated Tailwind CSS, so we need to run `npm run build:prod`. I'll do that before merging.

laymonage reviewed (approved):
Thanks! I was expecting we'd reuse the `components/card.html` template, but I suppose it's a bit more work to refactor that template and its usage to work with `CardBlock`.

I'll merge this for now along with the minor fixes I left in the comments, but I think it'd be good to get some design input on what the expected look for the block is (in relation to the existing `card.html`).

vossisboss commented:
Yeah, I missed the existing card template.  I found it when I was looking into a fix for #49 and I think that structure is okay but might be too repetitive design-wise. We should definitely speak to someone with better design taste than mine though.
